From 229b98ca2239bdf64cee8688104e22d2fac7367a Mon Sep 17 00:00:00 2001
From: JW Wang <wangchun@google.com>
Date: Mon, 3 Jan 2022 15:43:54 +0800
Subject: [PATCH 78/82] [Q_asb_2022-03] Don't abandon child sessions (1/n)

It will throw if abandon() is called on a child session.

Bug: 211944991
Bug: 67862680
Test: to be added
Change-Id: Ib0ba9f3786dda2d3174f3ea8c65d1061a3fcb586
Merged-In: Ib0ba9f3786dda2d3174f3ea8c65d1061a3fcb586
(cherry picked from commit 8b67e7db79d15ab448ae8f00b40a4a350ab3d537)
(cherry picked from commit a4ef9e0e00416d8e211bd749da42033ae4fd9eb2)
Merged-In: Ib0ba9f3786dda2d3174f3ea8c65d1061a3fcb586
---
 .../com/android/server/pm/PackageInstallerService.java    | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/services/core/java/com/android/server/pm/PackageInstallerService.java b/services/core/java/com/android/server/pm/PackageInstallerService.java
index e7b0df397f54..0259d48bf488 100644
--- a/services/core/java/com/android/server/pm/PackageInstallerService.java
+++ b/services/core/java/com/android/server/pm/PackageInstallerService.java
@@ -337,7 +337,13 @@ public class PackageInstallerService extends IPackageInstaller.Stub implements
                 if (age >= MAX_SESSION_AGE_ON_LOW_STORAGE_MILLIS) {
                     // Aggressively close old sessions because we are running low on storage
                     // Their staging dirs will be removed too
-                    session.abandon();
+                    PackageInstallerSession root = !session.hasParentSessionId()
+                            ? session : mSessions.get(session.getParentSessionId());
+                    if (!root.isDestroyed() && 
+                            (!root.isStaged() || (root.isStaged() && root.isStagedSessionReady()))) 
+                    {
+                        root.abandon();
+                    }
                 } else {
                     // Session is new enough, so it deserves to be kept even on low storage
                     unclaimedStagingDirsOnVolume.remove(session.stageDir);
-- 
2.35.1

